hazelcast/src/main/java/com/hazelcast/instance/HazelcastInstanceFactory.java

```diff
@@ -162,7 +162,28 @@ public static HazelcastInstance newHazelcastInstance(Config config) {
     }
 
     public static String createInstanceName(Config config) {
-        return "_hzInstance_" + FACTORY_ID_GEN.incrementAndGet() + "_" + config.getGroupConfig().getName();
+        Boolean useMobyNaming = Boolean.valueOf(config.getProperty(GroupProperty.MOBY_NAMING_ENABLED.getName()));
+        int instanceNum = FACTORY_ID_GEN.incrementAndGet();
+        String name;
+        if (useMobyNaming) {
+            name = createUniqueMobyName(instanceNum);
+        } else {
+            name = "_hzInstance_" + instanceNum + "_" + config.getGroupConfig().getName();
+        }
+        return name;
+    }
+
+    /**
+     * Generate random Moby Name. If a name already exists so {@param instanceNum} is appended to the end to provide uniqueness.
+     * @param instanceNum instance number. Must be unique within the current classloader.
+     * @return new instance's generated name.
+     */
+    private static String createUniqueMobyName(int instanceNum) {
+        String nameCandidate = MobyNames.getRandomName();
+        if (!INSTANCE_MAP.containsKey(nameCandidate)) {
+            return nameCandidate;
+        }
+        return nameCandidate + instanceNum;
```

Comment:
I think we should try again instead of appending instanceNum here. If you ever hit this line, you will have two very similarly named instances in same classloader such as `brave_driscoll` and `brave_driscoll2`. I think it opposes the goal of moby naming.

Reply:
Ok, I added several attempts but anyway there is a very little probability that after any amount of attempts we will still no succeed to choose unique name and we will have to make it by adding index at the end.